File: backtracking/java/leetcode46_Permutations.java
//Given a collection of distinct integers, return all possible permutations.
//
//Example:
//
//Input: [1,2,3]
//Output:
//[
//  [1,2,3],
//  [1,3,2],
//  [2,1,3],
//  [2,3,1],
//  [3,1,2],
//  [3,2,1]
//]
//

package wenxinjie;

import java.util.ArrayList;
import java.util.List;

public class leetcode46_Permutations {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        helper(res, new ArrayList<>(), nums);
        return res;
    }
    
    public void helper(List<List<Integer>> res, List<Integer> cur, int[] nums){
        if (cur.size() == nums.length)  res.add(new ArrayList<>(cur));
        for (int i: nums){
            if (cur.contains(i)) continue;
            cur.add(i);
            helper(res, cur, nums);
            cur.remove(cur.size()-1);              
        }
    }
	
}

//Time: O(n!)
//Space: O(n * n!)
//Difficulty: medium

File: backtracking/python/leetcode46_Permutations.py
# Given a collection of distinct integers, return all possible permutations.

# Example:

# Input: [1,2,3]
# Output:
# [
#   [1,2,3],
#   [1,3,2],
#   [2,1,3],
#   [2,3,1],
#   [3,1,2],
#   [3,2,1]
# ]

class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        self.helper(res, [], nums)
        return res
    
    def helper(self, res, cur, nums):
        if len(cur) == len(nums):
            res.append(cur + [])
            return
        for i in nums:
            if i in cur:
                continue
            cur.append(i)
            self.helper(res, cur, nums)
            cur.pop()


# Time: O(n!)
# Space: O(n * n!)
# Difficulty: medium
